packages/protocol/test/governance/governance.ts

```diff
@@ -1872,6 +1921,173 @@ contract('Governance', (accounts: string[]) => {
+      await governance.whitelistHotfix(proposalHashStr, { from: accounts[2] })
+      const passing = await governance.isHotfixPassing.call(proposalHashStr)
+      assert.isFalse(passing)
+    })
+
+    it('should return true when hotfix is whitelisted by quorum', async () => {
+      await governance.whitelistHotfix(proposalHashStr, { from: accounts[2] })
+      await governance.whitelistHotfix(proposalHashStr, { from: accounts[3] })
+      const passing = await governance.isHotfixPassing.call(proposalHashStr)
+      assert.isTrue(passing)
+    })
+  })
+
+  // hard code in ganache
+  const EPOCH = 100
+
+  describe('#prepareHotfix()', () => {
+    beforeEach(async () => {
+      await mockValidators.setByzantineQuorumForCurrentSet(1)
+      await mockValidators.addValidator(accounts[2])
+    })
+
+    it('should revert when hotfix is not passing', async () => {
+      await assertRevert(governance.prepareHotfix(proposalHashStr))
+    })
+
+    describe('when hotfix is passing', () => {
+      let currEpoch = 1
+      beforeEach(async () => {
+        await mineBlocks(EPOCH, web3)
+        currEpoch += 1
+        await governance.whitelistHotfix(proposalHashStr, { from: accounts[2] })
+      })
+
+      it('should mark the hotfix record prepared epoch', async () => {
+        await governance.prepareHotfix(proposalHashStr)
+        const [, , preparedEpoch] = await governance.getHotfixRecord.call(proposalHashStr)
+        assertEqualBN(preparedEpoch, currEpoch)
+      })
+
+      it('should emit the HotfixPrepared event', async () => {
+        const resp = await governance.prepareHotfix(proposalHashStr)
+        assert.equal(resp.logs.length, 1)
+        const log = resp.logs[0]
+        assertLogMatches2(log, {
+          event: 'HotfixPrepared',
+          args: {
+            txHash: matchAny,
+            epoch: currEpoch,
+          },
+        })
+        assert.isTrue(Buffer.from(stripHexEncoding(log.args.txHash), 'hex').equals(proposalHash))
+      })
+
+      it('should revert when epoch == preparedEpoch', async () => {
+        await governance.prepareHotfix(proposalHashStr)
+        await assertRevert(governance.prepareHotfix(proposalHashStr))
+      })
+
+      it('should succeed for epoch != preparedEpoch', async () => {
+        await governance.prepareHotfix(proposalHashStr)
+        await mineBlocks(EPOCH, web3)
+        await governance.prepareHotfix(proposalHashStr)
+      })
+    })
+  })
+
+  describe('#executeHotfix()', () => {
+    const executeHotfixTx = () =>
+      governance.executeHotfix(
+        [transactionSuccess1.value],
+        [transactionSuccess1.destination],
+        // @ts-ignore bytes type
+        transactionSuccess1.data,
+        [transactionSuccess1.data.length]
+      )
+
+    it('should revert when hotfix not approved', async () => {
+      await assertRevert(executeHotfixTx())
+    })
+
+    it('should revert when hotfix not prepared for current epoch', async () => {
```

Comment:
Can you add a test that it fails when we prepare the hotfix but then move into a new epoch?